src/main/java/com/artipie/asto/ArtipieStorage.java

```diff
@@ -0,0 +1,249 @@
+     * New storage slice.
+     *
+     * @param clients HTTP clients
+     * @param remote Remote URI
+     */
+    public ArtipieStorage(final ClientSlices clients, final URI remote) {
+        this(new UriClientSlice(clients, remote));
+    }
+
+    /**
+     * Ctor.
+     *
+     * @param remote Remote slice
+     */
+    ArtipieStorage(final Slice remote) {
+        this.remote = remote;
+    }
+
+    /**
+     * Ctor.
+     *
+     * @param clients HTTP clients
+     * @param remote Remote URI
+     * @param auth Authenticator
+     */
+    public ArtipieStorage(final ClientSlices clients, final URI remote,
+        final Authenticator auth) {
+        this(
+            new AuthClientSlice(new UriClientSlice(clients, remote), auth)
+        );
+    }
+
+    @Override
+    public CompletableFuture<Boolean> exists(final Key key) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletableFuture<Collection<Key>> list(final Key prefix) {
+        final CompletableFuture<Collection<Key>> promise = new CompletableFuture<>();
+        this.remote.response(
+            new RequestLine(RqMethod.GET, ArtipieStorage.uri(prefix)).toString(),
+            new Headers.From("Accept", "application/json"),
+            Content.EMPTY
+        ).send(
+            (status, rsheaders, rsbody) -> {
+                final CompletableFuture<Void> term = new CompletableFuture<>();
+                if (status.success()) {
+                    new PublisherAs(
+                        Flowable.fromPublisher(rsbody)
+                            .doOnError(term::completeExceptionally)
+                            .doOnTerminate(() -> term.complete(null))
+                    ).string(StandardCharsets.UTF_8)
+                        .thenApply(s -> promise.complete(ArtipieStorage.parse(s)));
+                } else {
+                    final String msg = String.format(
+                        "Cannot get lists blobs contained in given path [prefix=%s, status=%s]",
+                        prefix,
+                        status
+                    );
+                    promise.completeExceptionally(
+                        new ArtipieIOException(msg)
+                    );
+                }
+                return term;
+            }
+        );
+        return promise;
+    }
+
+    @Override
+    public CompletableFuture<Void> save(final Key key, final Content content) {
+        return this.remote.response(
+            new RequestLine(RqMethod.PUT, ArtipieStorage.uri(key)).toString(),
+            new Headers.From(new ContentLength(content.size().get())),
+            content
+        ).send(
+            (status, rsheaders, rsbody) -> {
+                final CompletableFuture<Void> res;
+                if (status.success()) {
+                    res = CompletableFuture.allOf();
+                } else {
+                    res = new CompletableFuture<>();
+                    res.completeExceptionally(
```

Comment:
hm, the expression seems odd (`res` is not initialized here), I'd suggest to use `FailedFuture` from artipie/asto

Reply:
Sorry, I could not find the `FailedFuture` from artipie/asto so I decided to use `CompletableFuture.failedFuture` here, but it is available since java 9. Is it a problem?

Reply:
sorry, it's `FailedCompletionStage` not `FailedFuture`. Let's use it and make this project compatible with java 8

Reply:
ok